components/layout_2020/taffy/layout.rs

```diff
@@ -0,0 +1,620 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+use std::marker::PhantomData;
+
+use app_units::Au;
+use atomic_refcell::{AtomicRef, AtomicRefCell, AtomicRefMut};
+use style::properties::ComputedValues;
+use style::values::generics::length::{GenericLengthPercentageOrAuto, LengthPercentageOrAuto};
+use style::values::specified::align::AlignFlags;
+use style::values::specified::box_::DisplayInside;
+use style::Zero;
+use taffy::style_helpers::{TaffyMaxContent, TaffyMinContent};
+use taffy::MaybeMath;
+
+use super::{TaffyContainer, TaffyItemBox, TaffyItemBoxInner, TaffyStyloStyle};
+use crate::cell::ArcRefCell;
+use crate::context::LayoutContext;
+use crate::formatting_contexts::{Baselines, IndependentFormattingContext, IndependentLayout};
+use crate::fragment_tree::{BoxFragment, CollapsedBlockMargins, Fragment};
+use crate::geom::{
+    AuOrAuto, LogicalSides, LogicalVec2, PhysicalPoint, PhysicalRect, PhysicalSides, PhysicalSize,
+    Size,
+};
+use crate::positioned::{AbsolutelyPositionedBox, PositioningContext, PositioningContextLength};
+use crate::sizing::{ContentSizes, InlineContentSizesResult};
+use crate::style_ext::ComputedValuesExt;
+use crate::{ContainingBlock, IndefiniteContainingBlock};
+
+const DUMMY_NODE_ID: taffy::NodeId = taffy::NodeId::new(u64::MAX);
+
+fn resolve_content_size(constraint: taffy::AvailableSpace, content_sizes: ContentSizes) -> f32 {
+    match constraint {
+        taffy::AvailableSpace::Definite(limit) => {
+            let min = content_sizes.min_content.to_f32_px();
+            let max = content_sizes.max_content.to_f32_px();
+            limit.min(max).max(min)
```

Comment:
Could this use [`clamp`](https://doc.rust-lang.org/std/primitive.f32.html#method.clamp) instead of a `min()` and then `max()` call?

Reply:
It can't because `clamp` panics if `min > max` (see docs you linked above), but that's valid in CSS (min takes precedence over max).

Reply:
The min-content shouldn't be greater than the max-content, but yeah I wouldn't rely on that due to bugs.
Maybe shorter as `content_sizes.shrink_to_fit(Au::from_f32_px(limit)).to_f32_px()`, but either is fine.